google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -406,66 +359,138 @@ async def seconds_until_refresh(self) -> int:
 
-    async def _schedule_refresh(self, delay: Optional[int] = None) -> asyncio.Task:
-        """A coroutine that sleeps for the specified amount of time before
-        running _perform_refresh.
+        :type delay: int
+        :param delay
+            Time in seconds to sleep before running _perform_refresh.
 
         :rtype: asyncio.Task
-        :returns: A Task representing _get_instance_data.
+        :returns: A Task representing the scheduled _perform_refresh.
         """
 
-        if delay is None:
-            delay = await self.seconds_until_refresh()
-        try:
-            logger.debug("Entering sleep")
-            await asyncio.sleep(delay)
-        except asyncio.CancelledError as e:
-            logger.debug("Schedule refresh task cancelled.")
-            raise e
-        return await self._perform_refresh()
+        async def _refresh_task(
+            self: InstanceConnectionManager, delay: int
+        ) -> InstanceMetadata:
+            """
+            A coroutine that sleeps for the specified amount of time before
+            running _perform_refresh.
+            """
+            refresh_task: asyncio.Task
+            try:
+                logger.debug("Entering sleep")
+                if delay > 0:
+                    await asyncio.sleep(delay)
+                refresh_task = self._loop.create_task(self._perform_refresh())
+                await refresh_task
+            except asyncio.CancelledError as e:
+                logger.debug("Schedule refresh task cancelled.")
+                raise e
+            # bad refresh attempt
+            except Exception as e:
+                logger.exception(
+                    "An error occurred while performing refresh. "
+                    "Scheduling another refresh attempt immediately",
+                    exc_info=e,
+                )
+                # check if current metadata is valid, don't want to replace valid metadata with invalid one
+                try:
+                    instance_data = await self._current
+                    if instance_data.expiration < datetime.datetime.now():
+                        raise ExpiredInstanceMetadata(
+                            "Instance metadata is expired, performing new refresh."
+                        )
+                except Exception:
+                    # Current result is invalid, no-op
+                    logger.debug("Current instance data is invalid.")
+                    self._current = refresh_task
+                # schedule new refresh attempt immediately
+                self._next = self._schedule_refresh(0)
+            # if valid refresh, replace current with valid metadata and schedule next refresh
+            else:
+                self._current = refresh_task
+                # Ephemeral certificate expires in 1 hour, so we schedule a refresh to happen in 55 minutes.
+                delay = await self._seconds_until_refresh()
```

Comment:
Can we make `_seconds_until_refresh` either something on InstanceMetadata or a standalone function? I don't think it needs access to `self` here.

Reply:
I'm also not sure we need another context switch here - we should have the result of `instance_data` from the await of `refresh_task` earlier, which should be what we need to calculate the delay